memory and flash optimizations for 3.12

leptun opened the pull request.

wavexx committed:
Tune filament unloading to perform a small extrusion

Re-organize variant #define's related to related feedrates (without
changing old defaults) and add some extra comments

wavexx committed:
Do not extrude during unload in a filament runout

Introduce a new internal code 'R' to cleanly differentiate between
regular filament swap or runout.

wavexx committed:
Fix "Change filament" within a pause

wavexx committed:
Show "Print paused" after a change of filament in a paused state

wavexx committed:
Remove hard-coded X/Y jerk limits

wavexx committed:
Raise Z 5cm after stopping

Check if Z axis has been homed and, if so, raise Z at least 5cm after
stopping so that the steel sheet can be comfortably removed.

wavexx committed:
Allow all microstep resolutions for all axes

Panayiotis-git committed:
Ignore button long press if not allowed

wavexx committed:
Introduce lcd_show_yes_no_and_wait for future use

Split off yes/no lcd handling from
lcd_show_fullscreen_message_yes_no_and_wait_P into a separate function
in order to allow generic requests where the text prompt can change.

wavexx committed:
Show the axes invoved during crash detection

Explicitly show which axes are affected on the status line when a crash
has been detected.

When the crash recovery prompt is triggered, also show all axes which
have been affected during the detection interval (currently X, Y, or
both).

The timer detection has been simplified to match the MK3 optical
filament timeout handling, which makes more sense: a crash immediately
following a crash recovery prompt will trigger the prompt again. The
timer needs to properly expire in order to reset the crash count.

This helps crash detection tuning (and help users reporting issues with
crash detection) without having to monitor the serial line.

This currently abuses the MSG_CRASH_DETECTED message by prepending the
[X][Y] labels to the message, which is ok in english but might not
translate that well for all languages.

wavexx committed:
Unify fullscreen/multiscreen/yes-no prompt message handling

Modify lcd_show_multiscreen_message_two_choices_and_wait_P to also
handle single-screen or empty (no-clear) prompts, making other functions
redundant. Saves 76 bytes.

Change existing functions to simply call
lcd_show_multiscreen_message_two_choices_and_wait_P with the correct
arguments.

This changes the prompt of existing Yes/No messages: the previous prompt
would use the last two lines of the LCD, while the new prompt is using
just the last line of the LCD instead.

Translation do not require updates, since the Yes/No translation was
already the same in both implementations.

wavexx committed:
Allow choice positioning. Balance Yes/No messages.

Add an additional parameter to control the position of second choice
prompt position (while defaulting to the old).

This allows Yes/No prompts to be equally spaced.

leptun committed:
Remove unused `skip_debug_msg` flag

leptun committed:
Remove unused `skip_debug_msg` flag

leptun committed:
Tmc2130 use ShortTimer instead of custom timer implementation

leptun committed:
Tmc2130 use ShortTimer instead of custom timer implementation

leptun committed:
Remove unused Stepper.cpp variables

leptun committed:
Remove unused Stepper.cpp variables

leptun committed:
TimeNow && TimeSent

leptun committed:
TimeNow && TimeSent

leptun committed:
nIRsensorLastTime

leptun committed:
nIRsensorLastTime

leptun committed:
Adjust temperature runaway extruder count

leptun committed:
Adjust temperature runaway extruder count

leptun committed:
Change arduino IO to fastio where possible

leptun committed:
Change arduino IO to fastio where possible

leptun committed:
Remove maxlimit_status

leptun committed:
Remove maxlimit_status

leptun committed:
Fix build

leptun committed:
Fix build

leptun committed:
Disable PRUSA M28

leptun committed:
Disable PRUSA M28

leptun committed:
Sound enums as uint8_t

leptun committed:
Sound enums as uint8_t

leptun committed:
Disable unimplemented PID_ADD_EXTRUSION_RATE

leptun committed:
Disable unimplemented PID_ADD_EXTRUSION_RATE

leptun committed:
M42 optimization

leptun committed:
M42 optimization

leptun committed:
Fix M226

leptun committed:
Fix M226

leptun commented:
Still left to do:
- [x] Make WELCOME_MSG non-translated progmem (the translation was dropped many years ago)
- [ ] Remove_unused_variables if there are any left.
- [ ] change the maaaaany 2byte ints to uint8_t/int8_t where appropriate.
- [ ] Merge hotend and bed PID loops. Just like with thermal runaway where the bed is index 0 and the hotends are index>=1
- [x] Pack these flags into two bytes (endstops and endstops_old) and use bitmasks for access:
```
5d2     1       old_z_max_endstop
5d3     1       z_max_endstop
5d4     1       old_z_min_endstop
5d5     1       z_min_endstop
5d6     1       old_y_max_endstop
5d7     1       y_max_endstop
5d8     1       old_y_min_endstop
5d9     1       y_min_endstop
5da     1       old_x_max_endstop
5db     1       x_max_endstop
5dc     1       old_x_min_endstop
5dd     1       x_min_endstop
```
- [x] Wth is this? Let's get rid of it
```
25c     1       firstrun
```
- [x] These three can be merged.
```
4da     1       endstop_y_hit
4db     1       endstop_x_hit
endstop_z_hit
```
- [ ] Should be ifdefed out with SNMM flag:
```
21d     2       bowden_length
```
- [ ] These two look redundant. Should be investigated:
```
26b     4       max_pos
277     4       min_pos
```
- [x] Is this really needed?
```
519     4       endstops_trigsteps
```
- [ ] We should change the Timer.cpp/h to be more useful for cases like these:
```
283     4       safetytimer_inactive_time
287     4       stepper_inactive_time
2a1     4       prev_busy_signal_ms
57f     4       _usb_timer
5b9     4       mmu_last_request
5bd     4       mmu_last_response
6f7     4       lcd_next_update_millis
702     4       NcTime
814     4       max_inactive_time
14a1    4       extruder_autofan_last_check
1598    4       kicktime
159c    4       PingTime
```
- [ ] PID autotune should be improved. Creating a secondary main loop is stupidly dangerous and flash hungry. The state machines should be done better. With that, these global variables become redundant:
```
6a4     4       _Kd
6a8     4       _Ki
6ac     4       _Kp
6b0     1       pid_tuning_finished
```
- [ ] Considering all the memory corruption that can happen, I feel pretty unsafe to have this flag in SRAM. Maybe we can use some IO register for this bit (eg, `GPIOR0`. It is <=0x3F, so the bit manipulation instructions work on this)
```
18e7    1       bedPWMDisabled
```
- [ ] This flag can be removed since the timer already has a flag built-in:
```
bIRsensorStateFlag
```

gudnimg commented:
> * change the maaaaany 2byte ints to uint8_t/int8_t where appropriate.

I think all uses of `fanSpeedBckp` and `fanSpeed` can be changed from `int` to `uint8_t`. 

Especially changing `fanSpeed` to `uint8_t` in `Marlin.h` and `Marlin_main.cpp` saves `126` bytes of flash and `1` byte of dynamic memory on my end using the code in this PR.

Edit:
Changing this line to `uint8_t` saves additional `16` bytes of flash and `16` bytes of dynamic memory.
https://github.com/prusa3d/Prusa-Firmware/blob/a8a253c0ae115d9e4ec75adf62018436019a50d6/Firmware/planner.h#L108

leptun commented:
@gudnimg Yes. These fan speeds that are uint16_t can be changed all around the code. There are a lot more scenarios like this for other variables, so I just generalized all of them into this (from the todo list):
> - change the maaaaany 2byte ints to uint8_t/int8_t where appropriate.

DRracer committed:
Merge branch 'MK3' into 1.11_Optimizations

leptun committed:
Fix build

leptun committed:
Fix build

gudnimg commented:
Adding a comment here since I think its best to handle this in this PR

I noticed `backlightLevel_HIGH` and `backlightLevel_LOW` are declared as `int16_t` but we're storing them into memory as `uint8_t`.

```cpp
void backlight_save() //saves all backlight data to eeprom.
{
    eeprom_update_byte((uint8_t *)EEPROM_BACKLIGHT_LEVEL_HIGH, (uint8_t)backlightLevel_HIGH); // <--- uint8_t
    eeprom_update_byte((uint8_t *)EEPROM_BACKLIGHT_LEVEL_LOW, (uint8_t)backlightLevel_LOW); // <--- uint8_t
    eeprom_update_byte((uint8_t *)EEPROM_BACKLIGHT_MODE, backlightMode);
    eeprom_update_word((uint16_t *)EEPROM_BACKLIGHT_TIMEOUT, backlightTimer_period);
}

```
Changing these two variables to `uint8_t` saves `10` bytes of Flash and `2` bytes of dynamic memory.

Also `backlightTimer_period` should be declared `uint16_t` and not `int16_t` (for consistency).

wavexx committed:
Merge remote-tracking branch 'upstream/MK3' into show_axes_during_crash

wavexx committed:
Merge branch 'unify_fullscreen_msg_handling' into MK3

wavexx committed:
Merge branch 'show_axes_during_crash' into MK3

wavexx committed:
Merge branch 'tune_during_pause' into MK3

wavexx committed:
Merge branch 'remove_jerk_limits' into MK3

wavexx committed:
Merge remote-tracking branch 'panayiotis/MK3_ButtonLongPress_Fix' into MK3

wavexx committed:
Merge branch 'remove_mres_limits' into MK3

wavexx committed:
Merge branch 'enhanced_unload' into MK3

# Conflicts:
#	Firmware/Marlin_main.cpp

wavexx committed:
Merge branch 'stop_raise_higher' into MK3

gudnimg commented:
More potential savings:

`homeaxis()` function: change the parameter type of axis to `uint8_t`
```c
#ifdef TMC2130
void homeaxis(int axis, uint8_t cnt, uint8_t* pstep)
#else
void homeaxis(int axis, uint8_t cnt)
#endif //TMC2130
```

There was another function somewhere which just changing the type of the parameters saved a bit of flash. I can't remember where I found it, I will edit this comment if I find it again. 😃 

Edit:  found it

If we change 

```c
static void lcd_selftest_screen_step(int _row, int _col, int _state, const char *_name_PROGMEM, const char *_indicator)
```
to this:
```c
static void lcd_selftest_screen_step(uint8_t _row, uint8_t _col, uint8_t _state, const char *_name_PROGMEM, const char *_indicator)
```
We save 62 bytes of flash (at least on my end).

gudnimg commented:
Is there any way for me to help with this PR. Is it possible for me to create PRs for it? 😊

leptun commented:
Yes, of course. PRs to this one are welcome. I don’t think there can be multiple PRs for this in the main repo directly since the memory optimizations will be merged into MK3 at once, not in parts.
I’d probably start with the easy fixes, such as converting ints to uint8_ts wherever possible.
Something I can recommend is keeping a log of the compiled output size after each change. The code size usually goes down, but other times it becomes way larget for no reason. If you can’t make it smaller again, note it down so that it can be checked again at the end.

gudnimg committed:
Use uint8_t when using axis enum

Saves 76 bytes of flash memory

gudnimg committed:
Use memset instead of nested for-loop to zero a 2d array

Saves 26 bytes of flash memory and removes two 'int' types

gudnimg committed:
Improve lcd_selftest_screen_step() parameter types

Saves 62 bytes of flash memory

gudnimg committed:
1. Remove redundant variable 'inters'
2. 'gh' variable can be 1 byte instead of 2

This saves 26 bytes of flash memory

gudnimg committed:
Remove unused global variable 'chunkHead'

No change in memory on my end, but it is one less 'int' :)

gudnimg committed:
Various optimisation to change int to uint8_t

This commit saves 174 bytes of flash memory

gudnimg committed:
Change temp_runaway_status from float to uint8_t and its enum

Saves 90 bytes of flash, and 12 bytes of SRAM

gudnimg committed:
Change temp_runaway_timer from float to uint32_t

Saves 32 bytes of flash

Also change temp_runaway_error_counter from int to
uint16_t to be specific

gudnimg committed:
Merge remote-tracking branch 'upstream/1.11_Optimizations' into gudni-PR1

gudnimg committed:
Merge branch '1.11_Optimizations' into gudni-PR1

gudnimg committed:
Remove unused variable

Saves 2 bytes of SRAM

gudnimg committed:
Change two arrays from int to uint8_t

__preheat_counter has max value of 16
__preheat_errors has max value of 5

Saves 58 bytes of flash memory and 4 bytes of SRAM

gudnimg committed:
temp_runaway_check() parameter _heater_id should be uint8_t

Saves 58 bytes of flash

leptun committed:
Merge pull request #6 from gudnimg/gudni-PR1

Purged some `int`s for `uint8_t` when using the axis enum

leptun committed:
Merge branch 'MK3' into 1.11_Optimizations

leptun committed:
Fix LUT loop

leptun commented on Firmware/tmc2130.cpp:
```diff
@@ -963,7 +963,7 @@ void tmc2130_set_wave(uint8_t axis, uint8_t amp, uint8_t fac1000)
 			reg >>= 1;
 //		printf("%3d\t%3d\t%2d\t%2d\t%2d\t%2d    %08x\n", i, vA, dA, b, w[s], s, reg);
-	}
+	} while (i++ != 255);
```
@gudnimg please double check if it behaves correctly now

gudnimg replied:
Was the code not working before? 🤔 

I have a vague idea of what the code does but I don't see any issues. We start at `i = 0` and loop through `i = 255` (microstep number 256) which I believe your code will do.

leptun replied:
The old code would loop indefinitely since uint8_t is always smaller than 256.

gudnimg replied:
Ah I see, can't believe I missed that.

gudnimg committed:
Use memset to zero previous_speed array

Saves 150 bytes of flash memory

gudnimg committed:
Use inlined function set_destination_to_current() in more places

Saves 70 bytes of flash memory

gudnimg committed:
Replace for-loop with memcpy

Saves 22 bytes of flash memory

gudnimg committed:
Change row_offsets[] to uint8_t

Saves 10 bytes of flash, and 4 bytes of SRAM

gudnimg committed:
Change return types of two functions uint8_t

These functions aren't used but they should return the appropriate type.

gudnimg committed:
Replace three lines by one function call which does the same thing

Saves 42 bytes of flash memory

gudnimg committed:
setTargetHotend expect an uint8_t, not int for second parameter

also change _usb_timer to ShortTimer

Saves 28 bytes of Flash and 1 byte of SRAM

gudnimg committed:
Convert mmu_last_request and mmu_last_response to ShortTimer

Saves 180 bytes of flash and 2 byte of SRAM

gudnimg committed:
Convert previous_millis_cmd to LongTimer

Saves 196 bytes of Flash but adds 1 byte of SRAM

gudnimg committed:
Convert autostart_atmillis to ShortTimer

Saves 34 bytes of flash and 1 byte of SRAM

gudnimg committed:
Convert NcTime to ShortTimer

The variable is only used in ultralcd.cpp so I made it static there
and deleted it from Marlin_main.cpp

Saves 24 bytes of flash and 1 byte of SRAM

gudnimg committed:
Convert mmu_last_finda_response to ShortTimer

Saves 48 bytes of Flash, 1 byte of SRAM

gudnimg committed:
Convert display_time to ShortTimer

Saves 24 bytes of flash, 1 byte of SRAM

gudnimg committed:
crashDetTimer can be ShortTimer

Expired value is less than 65535

Saves 8 bytes of flash and 2 bytes of SRAM

gudnimg committed:
Remove unused kicktime timer

The variable is only set and never actually used

Saves 8 bytes of flash and 4  bytes of SRAM

gudnimg committed:
Remove internal Prusa command 'Beat'

It doesn't seem to do anything.

Saves 22 bytes of flash

gudnimg committed:
Change usb_printing_counter to uint8_t

It is assigned max value of 10

Saves 26 byte of flash and 1 byte of SRAM

gudnimg committed:
Change heating_status to uint8_t

Saves 66 byte of flash and 1 byte of SRAM

gudnimg committed:
Change heating_status_counter to uint8_t

It has a max value of 14

Saves 28 bytes of flash and 1 byte of SRAM

gudnimg committed:
Change custom_message_state_old  and custom_message_state to uint8_t

I see max value as 7*7+10 = 59

Saves 100 byte of flash and 1 byte of SRAM

gudnimg committed:
Change farm_timer to uint8_t

Remove extern declaration of farm_timer and farm_status as they
are only used in one file.

Saves 26 bytes of flash

gudnimg committed:
Create enum class for heating_status variable

gudnimg committed:
Remove variable move_menu_scale

The variable is just set to 1.0 and is never modified, I also
removed the function lcd_move_menu_1mm()

Changes save 96 bytes of flash memory and 4 bytes of SRAM

gudnimg committed:
Smaller code in G28

Replace two blocks of code with already defined functions.

Change saves 100 bytes of flash memory and 11 lines of code.

gudnimg committed:
G81: Change for loop variables x and y to uint8_t

Saves 198 bytes of flash, that's crazy

gudnimg committed:
Use code_value_short()  in get_command()

Same code but uses a defined function

Saves 28 bytes of flash memory

gudnimg committed:
Use code_value_uint8() in get_command()

The if statement is simplified as the value is never negative
and we can get rid of one variable

Saves 12 bytes of flash memory

leptun committed:
Merge pull request #7 from gudnimg/gudni-PR2

Save some memory by using memset and inlined memcpy

leptun commented:
Hmmm… there seems to be some issue with the commits. I don’t remember having those commits at the beginning.

scottmudge commented:
There is a bug in `Marlin_main.cpp`.

At line ~`5679`:

```
for (int y = MESH_NUM_Y_POINTS-1; y >= 0; y--) {
```

was changed to 

```
for (uint8_t y = MESH_NUM_Y_POINTS-1; y >= 0; y--) {
```

to save a couple dynamic bytes. However, the test/break condition `y >= 0` will never be satisfied (it will always be true) if y is declared unsigned. 

In other areas, similar loops use `int8_t`, which is what I assume you meant to use.

gudnimg commented:
> There is a bug in `Marlin_main.cpp`.
> 
> At line ~`5679`:
> 
> ```
> for (int y = MESH_NUM_Y_POINTS-1; y >= 0; y--) {
> ```
> 
> was changed to
> 
> ```
> for (uint8_t y = MESH_NUM_Y_POINTS-1; y >= 0; y--) {
> ```
> 
> to save a couple dynamic bytes. However, the test/break condition `y >= 0` will never be satisfied (it will always be true) if y is declared unsigned.
> 
> In other areas, similar loops use `int8_t`, which is what I assume you meant to use.

@scottmudge Yup you're right, good catch. 👍 

EDIT:
Interesting, changing this from `uint8_t` to `int8_t` increases flash usage by 212 bytes on my end.

gudnimg committed:
Remove one unused variable

Saves 1 byte of SRAM

The variable is only assigned a value but is never used.

gudnimg committed:
Fix an issue found in code review

Unfortunately this increases flash usages by 212 bytes (same usage as in current release) It seems it is most memory effcient to use int16_t (int). int8_t requires more memory.

gudnimg committed:
Change set_z() parameters to uin8_t

Saves 24 bytes of flash

gudnimg committed:
Remove undefined function

No change in memory footprint.

gudnimg committed:
Loop index can be unsigned

Saves 4 bytes of flash memory

leptun committed:
Fix G81 for loop

leptun committed:
Merge pull request #8 from gudnimg/remove-unused-var-gudni

Remove unused variables
Shrink integer types where possible

leptun commented:
Closing in favor of #3365
